### PR TITLE
feat: port hard constraints for type check from Draco1

### PR DESCRIPTION
**Pull request by Zehua-Zeng**
Hard constraints from Draco1 hard.lp "type check" section.

### CODE REVIEW

**Comment by domoritz** [comment]
Please update the title of the pull request according to https://www.conventionalcommits.org/en/v1.0.0/

**Comment by Zehua-Zeng** [comment]
> Please update the title of the pull request according to https://www.conventionalcommits.org/en/v1.0.0/. Also make sure that tests pass locally.

Done.

**Review by domoritz — CHANGES_REQUESTED** [review]
Do we still need these checks with https://github.com/cmudig/draco2/blob/0beb90705bbe6108e143c703b9c4c3db7df74f4b/draco/asp/constraints.lp#L4? I suspect that many checks are redundant now. Try removing them but leave the tests and see whether they start failing.

File: draco/asp/hard.lp [hunk 1]
```diff
@@ -1,46 +1,97 @@
...
-violation(count_with_field) :-
-    attribute((encoding,aggregate),E,count),
-    attribute((encoding,field),E,_).
-
-% @hard(point_tick_bar_without_x_or_y) Point, tick, and bar require x or y channel.
-violation(point_tick_bar_without_x_or_y) :-
-    attribute((mark,type),M,(point;tick;bar)),
-    not attribute(mark_channel,M,x),
-    not attribute(mark_channel,M,y).
+% ====== Hard constraints ======
```
Comment (domoritz) [inline draco/asp/hard.lp:1]: Why has this changed?

File: draco/asp/helpers.lp [hunk 2]
```diff
@@ -1,6 +1,6 @@
-% ====== Helpers ======
-
-% @helper(mark_channel) Whether a mark has an encoding for a channel.
-attribute(mark_channel,M,C) :-
-    entity(encoding,M,E),
-    attribute((encoding,channel),E,C).
+% ====== Helpers ======
```
Comment (domoritz) [inline draco/asp/helpers.lp:1]: Something changed about the whitespaced here. Please undo.

**Review by domoritz — CHANGES_REQUESTED** [review]

**Comment by domoritz** [comment]
You don't need to reopen a new pull request. You can usually push code to the branch for a pull request.

**Comment by Zehua-Zeng** [comment]
I tried to merge the new version but did something wrong, and messed up the code, so I deleted the branch and recreated one. That's why this pull request was closed. Please review the new pull request. Thank you :)

**Pull request closed without merging**